test/tap/anon-cli-metrics.js

```diff
@@ -0,0 +1,148 @@
+'use strict'
+var path = require('path')
+var fs = require('graceful-fs')
+var test = require('tap').test
+var mr = require('npm-registry-mock')
+var Tacks = require('tacks')
+var File = Tacks.File
+var Dir = Tacks.Dir
+var extend = Object.assign || require('util')._extend
+var common = require('../common-tap.js')
+
+var basedir = path.join(__dirname, path.basename(__filename, '.js'))
+var testdir = path.join(basedir, 'testdir')
+var cachedir = path.join(basedir, 'cache')
+var globaldir = path.join(basedir, 'global')
+var tmpdir = path.join(basedir, 'tmp')
+var statsFile = path.join(cachedir, 'anonymous-cli-stats.json')
+
+var conf = {
+  cwd: testdir,
+  env: extend(extend({}, process.env), {
+    npm_config_cache: cachedir,
+    npm_config_tmp: tmpdir,
+    npm_config_prefix: globaldir,
+    npm_config_registry: common.registry,
+    npm_config_metrics_registry: common.registry,
+    npm_config_loglevel: 'warn'
+  })
+}
+
+var server
+var fixture = new Tacks(Dir({
+  cache: Dir(),
+  global: Dir(),
+  tmp: Dir(),
+  testdir: Dir({
+    failure: Dir({
+      'package.json': File({
+        name: 'failure',
+        version: '1.0.0',
+        scripts: {
+          preinstall: 'false'
+        }
+      })
+    }),
+    success: Dir({
+      'package.json': File({
+        name: 'success',
+        version: '1.0.0'
+      })
+    }),
+    slow: Dir({
+      'package.json': File({
+        name: 'slow',
+        version: '1.0.0',
+        scripts: {
+          preinstall: "node -e 'setTimeout(function(){}, 1500)'"
+        }
+      })
+    }),
+    'package.json': File({
+      name: 'anon-cli-metrics-test',
+      version: '1.0.0'
+    })
+  })
+}))
+
+function setup () {
+  cleanup()
+  fixture.create(basedir)
+}
+
+function cleanup () {
+  fixture.remove(basedir)
+}
+
+test('setup', function (t) {
+  setup()
+  mr({port: common.port, throwOnUnmatched: true}, function (err, s) {
+    if (err) throw err
+    server = s
+    server.filteringPathRegEx(/([/]-[/]npm[/]anon-metrics[/]v1[/]).*/, '$1:id')
+    server.filteringRequestBody(function (body) {
+      var metrics = typeof body === 'string' ? JSON.parse(body) : body
+      delete metrics.from
+      delete metrics.to
+      return JSON.stringify(metrics)
+    })
+    t.done()
+  })
+})
+
+test('record success', function (t) {
+  common.npm(['install', '--no-send-metrics', 'success'], conf, function (err, code, stdout, stderr) {
```

Comment:
I like that the tests don't rely on specific knowledge of what the metrics file looks like. This is a good set of integration tests! But running three installs to check one piece of behavior feels like a lot of overhead for the results. I don't think whacking this (and the next) install to speed up the test is a blocker, but if it could be done without a huge amount of trouble, it would be beneficial.